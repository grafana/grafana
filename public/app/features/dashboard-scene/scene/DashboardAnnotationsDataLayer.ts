import {
  AnnotationEvent,
  arrayToDataFrame,
  DataTopic,
  getDefaultTimeRange,
  PanelData,
  LoadingState,
} from '@grafana/data';
import { config } from '@grafana/runtime';
import { dataLayers } from '@grafana/scenes';
import { AnnotationQuery } from '@grafana/schema';
import { PublicAnnotationsDataSource } from 'app/features/query/state/DashboardQueryRunner/PublicAnnotationsDataSource';

/**
 * This class is an extension to dataLayers.AnnotationsDataLayer to provide support for public dashboards.
 */
export class DashboardAnnotationsDataLayer extends dataLayers.AnnotationsDataLayer {
  protected async resolveDataSource(query: AnnotationQuery) {
    if (config.publicDashboardAccessToken) {
      return new PublicAnnotationsDataSource();
    }
    return super.resolveDataSource(query);
  }

  protected processEvents(
    query: AnnotationQuery,
    events: {
      state: LoadingState;
      events: AnnotationEvent[];
    }
  ) {
    if (config.publicDashboardAccessToken) {
      const stateUpdate: PanelData = {
        series: [],
        timeRange: getDefaultTimeRange(),
        state: events.state,
      };

      const df = arrayToDataFrame(events.events);
      df.meta = {
        ...df.meta,
        dataTopic: DataTopic.Annotations,
      };

      stateUpdate.annotations = [df];

      return stateUpdate;
    } else {
      return super.processEvents(query, events);
    }
  }
}
