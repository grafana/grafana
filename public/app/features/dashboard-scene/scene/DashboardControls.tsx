import { css } from '@emotion/css';

import { GrafanaTheme2, VariableHide } from '@grafana/data';
import { selectors } from '@grafana/e2e-selectors';
import {
  SceneObjectState,
  SceneObject,
  SceneObjectBase,
  SceneComponentProps,
  SceneTimePicker,
  SceneRefreshPicker,
  SceneDebugger,
  VariableDependencyConfig,
  sceneGraph,
  SceneObjectUrlSyncConfig,
  SceneObjectUrlValues,
  CancelActivationHandler,
} from '@grafana/scenes';
import { Box, Stack, useStyles2 } from '@grafana/ui';

import { PanelEditControls } from '../panel-edit/PanelEditControls';
import { getDashboardSceneFor } from '../utils/utils';

import { DashboardLinksControls } from './DashboardLinksControls';

export interface DashboardControlsState extends SceneObjectState {
  variableControls: SceneObject[];
  timePicker: SceneTimePicker;
  refreshPicker: SceneRefreshPicker;
  hideTimeControls?: boolean;
  hideVariableControls?: boolean;
  hideLinksControls?: boolean;
}

export class DashboardControls extends SceneObjectBase<DashboardControlsState> {
  static Component = DashboardControlsRenderer;

  protected _variableDependency = new VariableDependencyConfig(this, {
    onAnyVariableChanged: this._onAnyVariableChanged.bind(this),
  });

  protected _urlSync = new SceneObjectUrlSyncConfig(this, {
    keys: ['_dash.hideTimePicker', '_dash.hideVariables', '_dash.hideLinks'],
  });

  /**
   * We want the hideXX url keys to only sync one way (url => state) on init
   * We don't want these flags to be added to URL.
   */
  getUrlState() {
    return {};
  }

  updateFromUrl(values: SceneObjectUrlValues) {
    const { hideTimeControls, hideVariableControls, hideLinksControls } = this.state;
    const isEnabledViaUrl = (key: string) => values[key] === 'true' || values[key] === '';

    // Only allow hiding, never "unhiding" from url
    // Becasue this should really only change on first init it's fine to do multiple setState here

    if (!hideTimeControls && isEnabledViaUrl('_dash.hideTimePicker')) {
      this.setState({ hideTimeControls: true });
    }

    if (!hideVariableControls && isEnabledViaUrl('_dash.hideVariables')) {
      this.setState({ hideVariableControls: true });
    }

    if (!hideLinksControls && isEnabledViaUrl('_dash.hideLinks')) {
      this.setState({ hideLinksControls: true });
    }
  }

  public constructor(state: Partial<DashboardControlsState>) {
    super({
      variableControls: [],
      timePicker: state.timePicker ?? new SceneTimePicker({}),
      refreshPicker: state.refreshPicker ?? new SceneRefreshPicker({}),
      ...state,
    });

    this.addActivationHandler(() => {
      let refreshPickerDeactivation: CancelActivationHandler | undefined;

      if (this.state.hideTimeControls) {
        refreshPickerDeactivation = this.state.refreshPicker.activate();
      }

      return () => {
        if (refreshPickerDeactivation) {
          refreshPickerDeactivation();
        }
      };
    });
  }

  /**
   * Links can include all variables so we need to re-render when any change
   */
  private _onAnyVariableChanged(): void {
    const dashboard = getDashboardSceneFor(this);
    if (dashboard.state.links?.length > 0) {
      this.forceRender();
    }
  }

  public hasControls(): boolean {
    const hasVariables = sceneGraph
      .getVariables(this)
      ?.state.variables.some((v) => v.state.hide !== VariableHide.hideVariable);
    const hasAnnotations = sceneGraph.getDataLayers(this).some((d) => d.state.isEnabled && !d.state.isHidden);
    const hasLinks = getDashboardSceneFor(this).state.links?.length > 0;
    const hideLinks = this.state.hideLinksControls || !hasLinks;
    const hideVariables = this.state.hideVariableControls || (!hasAnnotations && !hasVariables);
    const hideTimePicker = this.state.hideTimeControls;

    return !(hideVariables && hideLinks && hideTimePicker);
  }
}

function DashboardControlsRenderer({ model }: SceneComponentProps<DashboardControls>) {
  const { variableControls, refreshPicker, timePicker, hideTimeControls, hideVariableControls, hideLinksControls } =
    model.useState();
  const dashboard = getDashboardSceneFor(model);
  const { links, editPanel } = dashboard.useState();
  const styles = useStyles2(getStyles);
  const showDebugger = location.search.includes('scene-debugger');

  if (!model.hasControls()) {
    return null;
  }

  return (
    <div data-testid={selectors.pages.Dashboard.Controls} className={styles.controls}>
      <Stack grow={1} wrap={'wrap'}>
        {!hideVariableControls && variableControls.map((c) => <c.Component model={c} key={c.state.key} />)}
        <Box grow={1} />
        {!hideLinksControls && !editPanel && <DashboardLinksControls links={links} uid={dashboard.state.uid} />}
        {editPanel && <PanelEditControls panelEditor={editPanel} />}
      </Stack>
      {!hideTimeControls && (
        <Stack justifyContent={'flex-end'}>
          <timePicker.Component model={timePicker} />
          <refreshPicker.Component model={refreshPicker} />
        </Stack>
      )}
      {showDebugger && <SceneDebugger scene={model} key={'scene-debugger'} />}
    </div>
  );
}

function getStyles(theme: GrafanaTheme2) {
  return {
    controls: css({
      display: 'flex',
      alignItems: 'flex-start',
      flex: '100%',
      gap: theme.spacing(1),
      flexDirection: 'row',
      flexWrap: 'nowrap',
      position: 'relative',
      width: '100%',
      marginLeft: 'auto',
      [theme.breakpoints.down('sm')]: {
        flexDirection: 'column-reverse',
        alignItems: 'stretch',
      },
    }),
    embedded: css({
      background: 'unset',
      position: 'unset',
    }),
  };
}
