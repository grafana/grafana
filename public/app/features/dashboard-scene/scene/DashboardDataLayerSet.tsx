import {
  SceneDataLayerProviderState,
  SceneDataLayerProvider,
  SceneDataLayerSetBase,
  SceneComponentProps,
} from '@grafana/scenes';

import { AlertStatesDataLayer } from './AlertStatesDataLayer';

export interface DashboardDataLayerSetState extends SceneDataLayerProviderState {
  alertStatesLayer?: AlertStatesDataLayer;
  annotationLayers: SceneDataLayerProvider[];
}

export class DashboardDataLayerSet
  extends SceneDataLayerSetBase<DashboardDataLayerSetState>
  implements SceneDataLayerProvider
{
  public constructor(state: Partial<DashboardDataLayerSetState>) {
    super({
      ...state,
      name: state.name ?? 'Data layers',
      annotationLayers: state.annotationLayers ?? [],
    });

    this.addActivationHandler(() => this._onActivate());
  }

  private _onActivate() {
    this._subs.add(
      this.subscribeToState((newState, oldState) => {
        if (newState.annotationLayers !== oldState.annotationLayers) {
          this.querySub?.unsubscribe();
          this.subscribeToAllLayers(this.getAllLayers());
        }
      })
    );

    this.subscribeToAllLayers(this.getAllLayers());

    return () => {
      this.querySub?.unsubscribe();
    };
  }

  public addAnnotationLayer(layer: SceneDataLayerProvider) {
    this.setState({ annotationLayers: [...this.state.annotationLayers, layer] });
  }

  private getAllLayers() {
    const layers = [...this.state.annotationLayers];

    if (this.state.alertStatesLayer) {
      layers.push(this.state.alertStatesLayer);
    }

    return layers;
  }

  public static Component = ({ model }: SceneComponentProps<DashboardDataLayerSet>) => {
    const { annotationLayers } = model.useState();

    return (
      <>
        {annotationLayers.map((layer) => (
          <layer.Component model={layer} key={layer.state.key} />
        ))}
      </>
    );
  };
}
