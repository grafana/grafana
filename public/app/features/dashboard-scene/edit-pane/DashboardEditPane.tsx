import { css } from '@emotion/css';
import { useEffect, useRef } from 'react';

import { GrafanaTheme2 } from '@grafana/data';
import { SceneObjectState, SceneObjectBase, SceneObject, SceneObjectRef } from '@grafana/scenes';
import { ToolbarButton, useStyles2 } from '@grafana/ui';

import { EditableDashboardElement, isEditableDashboardElement } from '../scene/types';
import { getDashboardSceneFor } from '../utils/utils';

import { ElementEditPane } from './ElementEditPane';

export interface DashboardEditPaneState extends SceneObjectState {
  selectedObject?: SceneObjectRef<SceneObject>;
}

export class DashboardEditPane extends SceneObjectBase<DashboardEditPaneState> {
  public selectObject(obj: SceneObject) {
    const currentSelection = this.state.selectedObject?.resolve();
    if (currentSelection === obj) {
      const dashboard = getDashboardSceneFor(this);
      this.setState({ selectedObject: dashboard.getRef() });
    } else {
      this.setState({ selectedObject: obj.getRef() });
    }
  }
}

export interface Props {
  editPane: DashboardEditPane;
  isCollapsed: boolean;
  onToggleCollapse: () => void;
}

/**
 * Making the EditPane rendering completely standalone (not using editPane.Component) in order to pass custom react props
 */
export function DashboardEditPaneRenderer({ editPane, isCollapsed, onToggleCollapse }: Props) {
  // Activate the edit pane
  useEffect(() => {
    if (!editPane.state.selectedObject) {
      const dashboard = getDashboardSceneFor(editPane);
      editPane.setState({ selectedObject: dashboard.getRef() });
    }
    editPane.activate();
  }, [editPane]);

  const { selectedObject } = editPane.useState();
  const styles = useStyles2(getStyles);
  const paneRef = useRef<HTMLDivElement>(null);

  if (!selectedObject) {
    return null;
  }

  if (isCollapsed) {
    return (
      <div className={styles.expandOptionsWrapper}>
        <ToolbarButton
          tooltip={'Open options pane'}
          icon={'arrow-to-right'}
          onClick={onToggleCollapse}
          variant="canvas"
          className={styles.rotate180}
          aria-label={'Open options pane'}
        />
      </div>
    );
  }

  const element = getEditableElementFor(selectedObject.resolve());

  return (
    <div className={styles.wrapper} ref={paneRef}>
      <ElementEditPane element={element} key={element.getTypeName()} />
    </div>
  );
}

function getEditableElementFor(obj: SceneObject): EditableDashboardElement {
  if (isEditableDashboardElement(obj)) {
    return obj;
  }

  for (const behavior of obj.state.$behaviors ?? []) {
    if (isEditableDashboardElement(behavior)) {
      return behavior;
    }
  }

  throw new Error("Can't find editable element for selected object");
}

function getStyles(theme: GrafanaTheme2) {
  return {
    wrapper: css({
      display: 'flex',
      flexDirection: 'column',
      flex: '1 1 0',
      overflow: 'auto',
    }),
    rotate180: css({
      rotate: '180deg',
    }),
    expandOptionsWrapper: css({
      display: 'flex',
      flexDirection: 'column',
      padding: theme.spacing(2, 1),
    }),
  };
}
