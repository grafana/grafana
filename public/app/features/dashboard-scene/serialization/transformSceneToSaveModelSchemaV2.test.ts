import { VariableRefresh } from '@grafana/data';
import { config } from '@grafana/runtime';
import {
  AdHocFiltersVariable,
  behaviors,
  ConstantVariable,
  CustomVariable,
  DataSourceVariable,
  GroupByVariable,
  IntervalVariable,
  QueryVariable,
  SceneGridLayout,
  SceneRefreshPicker,
  SceneTimePicker,
  SceneTimeRange,
  SceneVariableSet,
  TextBoxVariable,
  VizPanel,
} from '@grafana/scenes';
import {
  DashboardCursorSync as DashboardCursorSyncV1,
  VariableHide as VariableHideV1,
  VariableSort as VariableSortV1,
} from '@grafana/schema/dist/esm/index.gen';

import { DashboardEditPane } from '../edit-pane/DashboardEditPane';
import { DashboardAnnotationsDataLayer } from '../scene/DashboardAnnotationsDataLayer';
import { DashboardControls } from '../scene/DashboardControls';
import { DashboardDataLayerSet } from '../scene/DashboardDataLayerSet';
import { DashboardScene, DashboardSceneState } from '../scene/DashboardScene';
import { VizPanelLinks, VizPanelLinksMenu } from '../scene/PanelLinks';
import { DashboardGridItem } from '../scene/layout-default/DashboardGridItem';
import { DefaultGridLayoutManager } from '../scene/layout-default/DefaultGridLayoutManager';

import { transformSceneToSaveModelSchemaV2 } from './transformSceneToSaveModelSchemaV2';

function setupDashboardScene(state: Partial<DashboardSceneState>): DashboardScene {
  return new DashboardScene(state);
}

jest.mock('@grafana/runtime', () => ({
  ...jest.requireActual('@grafana/runtime'),
  config: {
    ...jest.requireActual('@grafana/runtime').config,
    bootData: {
      settings: {
        defaultDatasource: 'loki',
        datasources: {
          Prometheus: {
            name: 'Prometheus',
            meta: { id: 'prometheus' },
            type: 'datasource',
          },
          '-- Grafana --': {
            name: 'Grafana',
            meta: { id: 'grafana' },
            type: 'datasource',
          },
          loki: {
            name: 'Loki',
            meta: { id: 'loki' },
            type: 'datasource',
          },
        },
      },
    },
  },
}));

describe('transformSceneToSaveModelSchemaV2', () => {
  let dashboardScene: DashboardScene;
  let prevFeatureToggleValue: boolean;

  beforeAll(() => {
    prevFeatureToggleValue = !!config.featureToggles.groupByVariable;
    config.featureToggles.groupByVariable = true;
  });

  afterAll(() => {
    config.featureToggles.groupByVariable = prevFeatureToggleValue;
  });

  beforeEach(() => {
    // The intention is to have a complete dashboard scene
    // with all the possible properties set
    dashboardScene = setupDashboardScene({
      $data: new DashboardDataLayerSet({ annotationLayers }),
      title: 'Test Dashboard',
      description: 'Test Description',
      preload: true,
      tags: ['tag1', 'tag2'],
      uid: 'test-uid',
      version: 1,
      $timeRange: new SceneTimeRange({
        timeZone: 'UTC',
        from: 'now-1h',
        to: 'now',
        weekStart: 'monday',
        fiscalYearStartMonth: 1,
        UNSAFE_nowDelay: '1m',
        refreshOnActivate: {
          afterMs: 10,
          percent: 0.1,
        },
      }),
      controls: new DashboardControls({
        refreshPicker: new SceneRefreshPicker({
          refresh: '5s',
          intervals: ['5s', '10s', '30s'],
          autoEnabled: true,
          autoMinInterval: '5s',
          autoValue: '5s',
          isOnCanvas: true,
          primary: true,
          withText: true,
          minRefreshInterval: '5s',
        }),
        timePicker: new SceneTimePicker({
          isOnCanvas: true,
          hidePicker: true,
        }),
      }),
      links: [
        {
          title: 'Test Link',
          url: 'http://test.com',
          asDropdown: false,
          icon: '',
          includeVars: false,
          keepTime: false,
          tags: [],
          targetBlank: false,
          tooltip: '',
          type: 'link',
        },
      ],
      body: new DefaultGridLayoutManager({
        grid: new SceneGridLayout({
          isLazy: false,
          children: [
            new DashboardGridItem({
              body: new VizPanel({
                key: 'test-panel-uid',
                pluginId: 'timeseries',
                title: 'Test Panel',
                titleItems: [
                  new VizPanelLinks({
                    rawLinks: [
                      { title: 'Test Link 1', url: 'http://test1.com', targetBlank: true },
                      { title: 'Test Link 2', url: 'http://test2.com' },
                    ],
                    menu: new VizPanelLinksMenu({}),
                  }),
                ],
                description: 'Test Description',
                hoverHeader: true,
                hoverHeaderOffset: 10,
                fieldConfig: { defaults: {}, overrides: [] },
                displayMode: 'transparent',
                pluginVersion: '7.0.0',
                $timeRange: new SceneTimeRange({
                  timeZone: 'UTC',
                  from: 'now-3h',
                  to: 'now',
                }),
              }),
              // Props related to repeatable panels
              // repeatedPanels?: VizPanel[],
              // variableName?: string,
              // itemHeight?: number,
              // repeatDirection?: RepeatDirection,
              // maxPerRow?: number,
            }),
          ],
        }),
      }),
      meta: {},
      editPane: new DashboardEditPane({}),
      $behaviors: [
        new behaviors.CursorSync({
          sync: DashboardCursorSyncV1.Crosshair,
        }),
        new behaviors.LiveNowTimer({
          enabled: true,
        }),
      ],
      $variables: new SceneVariableSet({
        // Test each of the variables
        variables: [
          new QueryVariable({
            name: 'queryVar',
            label: 'Query Variable',
            description: 'A query variable',
            skipUrlSync: false,
            hide: VariableHideV1.hideLabel,
            value: 'value1',
            text: 'text1',
            query: 'query1',
            definition: 'definition1',
            datasource: { uid: 'datasource1', type: 'prometheus' },
            sort: VariableSortV1.alphabeticalDesc,
            refresh: VariableRefresh.onDashboardLoad,
            regex: 'regex1',
            allValue: '*',
            includeAll: true,
            isMulti: true,
          }),
          new CustomVariable({
            name: 'customVar',
            label: 'Custom Variable',
            description: 'A custom variable',
            skipUrlSync: false,
            hide: VariableHideV1.dontHide,
            value: 'option1',
            text: 'option1',
            query: 'option1, option2',
            options: [
              { label: 'option1', value: 'option1' },
              { label: 'option2', value: 'option2' },
            ],
            isMulti: true,
            allValue: 'All',
            includeAll: true,
          }),
          new DataSourceVariable({
            name: 'datasourceVar',
            label: 'Datasource Variable',
            description: 'A datasource variable',
            skipUrlSync: false,
            hide: VariableHideV1.dontHide,
            value: 'value1',
            text: 'text1',
            regex: 'regex1',
            pluginId: 'datasource1',
            defaultOptionEnabled: true,
          }),
          new ConstantVariable({
            name: 'constantVar',
            label: 'Constant Variable',
            description: 'A constant variable',
            skipUrlSync: false,
            hide: VariableHideV1.dontHide,
            value: 'value4',
          }),
          new IntervalVariable({
            name: 'intervalVar',
            label: 'Interval Variable',
            description: 'An interval variable',
            skipUrlSync: false,
            hide: VariableHideV1.dontHide,
            value: '1m',
            intervals: ['1m', '5m', '10m'],
            autoEnabled: false,
            autoMinInterval: '1m',
            autoStepCount: 10,
          }),
          new TextBoxVariable({
            name: 'textVar',
            label: 'Text Variable',
            description: 'A text variable',
            skipUrlSync: false,
            hide: VariableHideV1.dontHide,
            value: 'value6',
          }),
          new GroupByVariable({
            name: 'groupByVar',
            label: 'Group By Variable',
            description: 'A group by variable',
            skipUrlSync: false,
            hide: VariableHideV1.dontHide,
            value: 'value7',
            text: 'text7',
            datasource: { uid: 'datasource2', type: 'prometheus' },
            defaultOptions: [
              { text: 'option1', value: 'option1' },
              { text: 'option2', value: 'option2' },
            ],
            isMulti: false,
            includeAll: false,
          }),
          new AdHocFiltersVariable({
            name: 'adhocVar',
            label: 'Adhoc Variable',
            description: 'An adhoc variable',
            skipUrlSync: false,
            hide: VariableHideV1.dontHide,
            datasource: { uid: 'datasource3', type: 'prometheus' },
            baseFilters: [
              {
                key: 'key1',
                operator: '=',
                value: 'value1',
                condition: 'AND',
              },
              {
                key: 'key2',
                operator: '=',
                value: 'value2',
                condition: 'OR',
              },
            ],
            filters: [
              {
                key: 'key3',
                operator: '=',
                value: 'value3',
                condition: 'AND',
              },
            ],
            defaultKeys: [
              {
                text: 'defaultKey1',
                value: 'defaultKey1',
                group: 'defaultGroup1',
                expandable: true,
              },
            ],
          }),
        ],
      }),
    });
  });

  it('should transform scene to save model schema v2', () => {
    const result = transformSceneToSaveModelSchemaV2(dashboardScene);
    expect(result).toMatchSnapshot();

    // Check that the annotation layers are correctly transformed
    expect(result.annotations).toHaveLength(3);
    // check annotation layer 3 with no datasource has the default datasource defined as type
    expect(result.annotations?.[2].spec.query.kind).toBe('loki');
  });
});

const annotationLayer1 = new DashboardAnnotationsDataLayer({
  key: 'layer1',
  query: {
    datasource: {
      type: 'grafana',
      uid: '-- Grafana --',
    },
    name: 'query1',
    enable: true,
    iconColor: 'red',
  },
  name: 'layer1',
  isEnabled: true,
  isHidden: false,
});

const annotationLayer2 = new DashboardAnnotationsDataLayer({
  key: 'layer2',
  query: {
    datasource: {
      type: 'prometheus',
      uid: 'abcdef',
    },
    name: 'query2',
    enable: true,
    iconColor: 'blue',
  },
  name: 'layer2',
  isEnabled: true,
  isHidden: true,
});

// this could happen if a dahboard was created from code and the datasource was not defined
const annotationLayer3NoDsDefined = new DashboardAnnotationsDataLayer({
  key: 'layer3',
  query: {
    name: 'query3',
    enable: true,
    iconColor: 'green',
  },
  name: 'layer3',
  isEnabled: true,
  isHidden: true,
});

const annotationLayers = [annotationLayer1, annotationLayer2, annotationLayer3NoDsDefined];
