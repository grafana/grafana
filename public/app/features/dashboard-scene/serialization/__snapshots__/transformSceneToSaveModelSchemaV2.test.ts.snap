// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformSceneToSaveModelSchemaV2 should transform scene to save model schema v2 1`] = `
{
  "annotations": [
    {
      "kind": "AnnotationQuery",
      "spec": {
        "builtIn": false,
        "datasource": {
          "type": "grafana",
          "uid": "-- Grafana --",
        },
        "enable": true,
        "filter": {
          "exclude": false,
          "ids": [],
        },
        "hide": false,
        "iconColor": "red",
        "name": "query1",
        "query": {
          "kind": "grafana",
          "spec": {
            "enable": true,
            "iconColor": "red",
            "name": "query1",
          },
        },
      },
    },
    {
      "kind": "AnnotationQuery",
      "spec": {
        "builtIn": false,
        "datasource": {
          "type": "prometheus",
          "uid": "abcdef",
        },
        "enable": true,
        "filter": {
          "exclude": false,
          "ids": [],
        },
        "hide": true,
        "iconColor": "blue",
        "name": "query2",
        "query": {
          "kind": "prometheus",
          "spec": {
            "enable": true,
            "iconColor": "blue",
            "name": "query2",
          },
        },
      },
    },
    {
      "kind": "AnnotationQuery",
      "spec": {
        "builtIn": false,
        "datasource": {
          "type": "loki",
          "uid": "Loki",
        },
        "enable": true,
        "filter": {
          "exclude": false,
          "ids": [],
        },
        "hide": true,
        "iconColor": "green",
        "name": "query3",
        "query": {
          "kind": "loki",
          "spec": {
            "enable": true,
            "iconColor": "green",
            "name": "query3",
          },
        },
      },
    },
  ],
  "cursorSync": "Crosshair",
  "description": "Test Description",
  "editable": true,
  "elements": {
    "test-panel-uid": {
      "kind": "Panel",
      "spec": {
        "data": {
          "kind": "QueryGroup",
          "spec": {
            "queries": [],
            "queryOptions": {},
            "transformations": [],
          },
        },
        "description": "Test Description",
        "links": [
          {
            "targetBlank": true,
            "title": "Test Link 1",
            "url": "http://test1.com",
          },
          {
            "title": "Test Link 2",
            "url": "http://test2.com",
          },
        ],
        "title": "Test Panel",
        "uid": "test-panel-uid",
        "vizConfig": {
          "kind": "timeseries",
          "spec": {
            "fieldConfig": {
              "defaults": {},
              "overrides": [],
            },
            "options": {},
            "pluginVersion": "7.0.0",
          },
        },
      },
    },
  },
  "layout": {
    "kind": "GridLayout",
    "spec": {
      "items": [
        {
          "kind": "GridLayoutItem",
          "spec": {
            "element": {
              "kind": "ElementReference",
              "name": "test-panel-uid",
            },
            "height": 0,
            "width": 0,
            "x": 0,
            "y": 0,
          },
        },
      ],
    },
  },
  "links": [
    {
      "asDropdown": false,
      "icon": "",
      "includeVars": false,
      "keepTime": false,
      "tags": [],
      "targetBlank": false,
      "title": "Test Link",
      "tooltip": "",
      "type": "link",
      "url": "http://test.com",
    },
  ],
  "liveNow": true,
  "preload": true,
  "schemaVersion": 40,
  "tags": [
    "tag1",
    "tag2",
  ],
  "timeSettings": {
    "autoRefresh": "5s",
    "autoRefreshIntervals": [
      "5s",
      "10s",
      "30s",
    ],
    "fiscalYearStartMonth": 1,
    "from": "now-1h",
    "hideTimepicker": false,
    "nowDelay": "1m",
    "quickRanges": [],
    "timezone": "UTC",
    "to": "now",
    "weekStart": "monday",
  },
  "title": "Test Dashboard",
  "variables": [
    {
      "kind": "QueryVariable",
      "spec": {
        "allValue": "*",
        "current": {
          "text": "text1",
          "value": "value1",
        },
        "datasource": {
          "type": "prometheus",
          "uid": "datasource1",
        },
        "definition": "definition1",
        "description": "A query variable",
        "hide": "hideLabel",
        "includeAll": true,
        "label": "Query Variable",
        "multi": true,
        "name": "queryVar",
        "options": [],
        "query": "query1",
        "refresh": "onDashboardLoad",
        "regex": "regex1",
        "skipUrlSync": false,
        "sort": "alphabeticalDesc",
      },
    },
    {
      "kind": "CustomVariable",
      "spec": {
        "allValue": "All",
        "current": {
          "text": "option1",
          "value": "option1",
        },
        "description": "A custom variable",
        "hide": "dontHide",
        "includeAll": true,
        "label": "Custom Variable",
        "multi": true,
        "name": "customVar",
        "options": [
          {
            "selected": true,
            "text": "option1",
            "value": "option1",
          },
          {
            "selected": false,
            "text": "option2",
            "value": "option2",
          },
        ],
        "query": "option1, option2",
        "skipUrlSync": false,
      },
    },
    {
      "kind": "DatasourceVariable",
      "spec": {
        "allValue": undefined,
        "current": {
          "text": "text1",
          "value": "value1",
        },
        "defaultOptionEnabled": true,
        "description": "A datasource variable",
        "hide": "dontHide",
        "includeAll": false,
        "label": "Datasource Variable",
        "multi": false,
        "name": "datasourceVar",
        "options": [],
        "pluginId": "datasource1",
        "refresh": "onDashboardLoad",
        "regex": "regex1",
        "skipUrlSync": false,
      },
    },
    {
      "kind": "ConstantVariable",
      "spec": {
        "current": {
          "text": "value4",
          "value": "value4",
        },
        "description": "A constant variable",
        "hide": "dontHide",
        "label": "Constant Variable",
        "name": "constantVar",
        "query": "value4",
        "skipUrlSync": true,
      },
    },
    {
      "kind": "IntervalVariable",
      "spec": {
        "auto": false,
        "auto_count": 10,
        "auto_min": "1m",
        "current": {
          "text": "1m",
          "value": "1m",
        },
        "description": "An interval variable",
        "hide": "dontHide",
        "label": "Interval Variable",
        "name": "intervalVar",
        "options": [
          {
            "selected": true,
            "text": "1m",
            "value": "1m",
          },
          {
            "selected": false,
            "text": "5m",
            "value": "5m",
          },
          {
            "selected": false,
            "text": "10m",
            "value": "10m",
          },
        ],
        "query": "1m,5m,10m",
        "refresh": "onTimeRangeChanged",
        "skipUrlSync": false,
      },
    },
    {
      "kind": "TextVariable",
      "spec": {
        "current": {
          "text": "value6",
          "value": "value6",
        },
        "description": "A text variable",
        "hide": "dontHide",
        "label": "Text Variable",
        "name": "textVar",
        "query": "value6",
        "skipUrlSync": false,
      },
    },
    {
      "kind": "GroupByVariable",
      "spec": {
        "current": {
          "text": "text7",
          "value": "value7",
        },
        "datasource": {
          "type": "prometheus",
          "uid": "datasource2",
        },
        "description": "A group by variable",
        "hide": "dontHide",
        "includeAll": false,
        "label": "Group By Variable",
        "multi": false,
        "name": "groupByVar",
        "options": [
          {
            "text": "option1",
            "value": "option1",
          },
          {
            "text": "option2",
            "value": "option2",
          },
        ],
        "skipUrlSync": false,
      },
    },
    {
      "kind": "AdhocVariable",
      "spec": {
        "baseFilters": [
          {
            "condition": "AND",
            "key": "key1",
            "operator": "=",
            "value": "value1",
          },
          {
            "condition": "OR",
            "key": "key2",
            "operator": "=",
            "value": "value2",
          },
        ],
        "datasource": {
          "type": "prometheus",
          "uid": "datasource3",
        },
        "defaultKeys": [
          {
            "expandable": true,
            "group": "defaultGroup1",
            "text": "defaultKey1",
            "value": "defaultKey1",
          },
        ],
        "description": "An adhoc variable",
        "filters": [
          {
            "condition": "AND",
            "key": "key3",
            "operator": "=",
            "value": "value3",
          },
        ],
        "hide": "dontHide",
        "label": "Adhoc Variable",
        "name": "adhocVar",
        "skipUrlSync": false,
      },
    },
  ],
}
`;
